### ISSUE TITLE
Do not prompt for Tomb password if secret key material doesn't need one

### ISSUE THREAD

**Issue by alphazo** [dyne/Tomb#179]
There is one little last issue with the GnuPG setup. When I open the Tomb container with a GnuPG encrypted keyfile I first get prompted for a Tomb password for which I can type anything since it no longer applies and then I get the GnuPG password prompt.
Would be nice if Tomb first tried to use the GnuPG _unprotected_ secret instead of prompting for a password?
## Generating a 100MB empty tomb

```
# tomb dig -s 100 test.tomb
 tomb  .  Commanded to dig tomb test.tomb
 tomb (*) Creating a new tomb in ./test.tomb
 tomb  .  Generating test.tomb of 100MiB
 100+0 records in
 100+0 records out
 104857600 bytes (105 MB) copied, 8,55695 s, 12,3 MB/s
 -rw------- 1 alpha users 100M  7 déc.  12:20 ./test.tomb
 tomb (*) Done digging test
 tomb  .  Your tomb is not yet ready, you need to forge a key and lock it:
 tomb  .  tomb forge ./test.tomb.key
 tomb  .  tomb lock ./test.tomb -k ./test.tomb.key
```
## Creating the encryption key and using a dummy temporary password to protect it

Password in that case is "weak".

```
 tomb forge test.tomb.key
 tomb  .  Commanded to forge key test.tomb.key
 tomb [W] An active swap partition is detected, this poses security risks.
 tomb (*) All your swaps are belong to crypt.  Good.
 tomb  .  Commanded to forge key test.tomb.key with cipher algorithm AES256
 tomb  .  This operation takes time, keep using this computer on other tasks,
 tomb  .  once done you will be asked to choose a password for your tomb.
 tomb  .  To make it faster you can move the mouse around.
 tomb  .  If you are on a server, you can use an Entropy Generation Daemon.
 256+0 records in
 256+0 records out
 256 bytes (256 B) copied, 111,361 s, 0,0 kB/s
 tomb (*) Choose the  password of your key: test.tomb.key
 tomb  .  (You can also change it later using 'tomb passwd'.)
 [GNUPG:] PROGRESS need_entropy X 8 16
 [GNUPG:] PROGRESS need_entropy X 16 16
 [GNUPG:] NEED_PASSPHRASE_SYM 9 3 2
 [GNUPG:] BEGIN_ENCRYPTION 2 9
 [GNUPG:] END_ENCRYPTION
 tomb  .  Key is valid.
 tomb  .  Done forging test.tomb.key
 tomb (*) Your key is ready:
 -rw------- 1 alpha users 531  7 déc.  12:26 test.tomb.key
```
## Encrypt and format container with previously generated key

```
 # tomb lock test.tomb -k test.tomb.key
 tomb  .  Commanded to lock tomb test.tomb
 tomb  .  Checking if the tomb is empty (we never step on somebody else's bones).
 tomb  .  Fine, this tomb seems empty.
 tomb  .  Key is valid.
 tomb  .  Locking using cipher: aes-xts-plain64:sha256
 tomb  .  A password is required to use key test.tomb.key
 tomb  .  Password OK.
 tomb (*) Locking test.tomb with test.tomb.key
 tomb  .  Formatting Luks mapped device.
 tomb  .  Formatting your Tomb with Ext3/Ext4 filesystem.
 tomb  .  Done locking test using Luks dm-crypt aes-xts-plain64:sha256
 tomb (*) Your tomb is ready in ./test.tomb and secured with key test.tomb.key
```
## Open the container with 'weak' password:

```
 # tomb open test.tomb -k test.tomb.key
 tomb  .  Commanded to open tomb test.tomb
 tomb [W] An active swap partition is detected, this poses security risks.
 tomb (*) All your swaps are belong to crypt.  Good.
 tomb  .  Valid tomb file found: ./test.tomb
 tomb  .  Key is valid.
 tomb  .  Mountpoint not specified, using default: /media/test.tomb
 tomb (*) Opening test.tomb on /media/test.tomb
 tomb  .  This tomb is a valid LUKS encrypted device.
 tomb  .  Cipher is "aes" mode "xts-plain64:sha256" hash "sha1"
 tomb  .  A password is required to use key test.tomb.key
 tomb  .  Password OK.
 tomb (*) Success unlocking tomb test
 tomb  .  Checking filesystem via /dev/loop0
 fsck from util-linux 2.25.2
 test: clean, 11/25168 files, 8848/100352 blocks
 tomb (*) Success opening test.tomb on /media/test.tomb
```
## Bury something in the tomb

```
 # touch /media/test.tomb/mydirtylittlesecrets.txt
```
## Close tomb

```
 # tomb close
```
## Encrypt key material with GnuPG key

```
 # gpg -d test.tomb.key | gpg -e -r 0x242D4DFB --armor -o test.tomb.key.gpg
 gpg: AES256 encrypted data
 gpg: 0x00D7B31FCBF93DD2: There is no assurance this key belongs to the named user
 pub  elg2048/0x00D7B31FCBF93DD2 2009-08-20 Alphazo <alphazo@gmail.com>
  Primary key fingerprint: D521 F47D 3A10 41E5 71A1  E4CD 6338 F4BC 242D 4DFB
       Subkey fingerprint: 14E0 A6FB 37C8 24B4 4745  5FE0 00D7 B31F CBF9 3DD2
  gpg: encrypted with 1 passphrase
```
## Time to securely wipe the password encrypted original key material

```
 # wipe test.tomb.key
```
## Open the tomb using GnuPG password protected key

**The issue comes here since Tomb will ask for a 'tomb' password but in fact anything can be entered at this point since the original password is gone.**
A second window will pop-up asking for the GnuPG key password and container will be mounted.

```
 # tomb open test.tomb -k test.tomb.key.gpg
 tomb  .  Commanded to open tomb test.tomb
 tomb [W] An active swap partition is detected, this poses security risks.
 tomb (*) All your swaps are belong to crypt.  Good.
 tomb  .  Valid tomb file found: ./test.tomb
 tomb  .  Key is valid.
 tomb  .  Mountpoint not specified, using default: /media/test.tomb
 tomb (*) Opening test.tomb on /media/test.tomb
 tomb  .  This tomb is a valid LUKS encrypted device.
 tomb  .  Cipher is "aes" mode "xts-plain64:sha256" hash "sha1"
 tomb  .  A password is required to use key test.tomb.key.gpg
 tomb  .  Password OK.
 tomb (*) Success unlocking tomb test
 tomb  .  Checking filesystem via /dev/loop0
 fsck from util-linux 2.25.2
 test: clean, 16/25168 files, 8852/100352 blocks
 tomb (*) Success opening test.tomb on /media/test.tomb
 tomb  .  Last visit by alpha(1000) from /dev/pts/0 on fatfly
 tomb  .  on date dim. 07 déc. 2014 12:28:48 CET
```
## Check if secret files are here

```
 # ls /media/test.tomb
 lost+found  mydirtylittlesecrets.txt
```
## Want something cool?

Minus the little annoyance with the extra password prompt the above setup also works nicely with a Cryptostick GnuPG drive that holds the GnuPG decryption key (instead of being stored on your PC). It provides an extra level of security and allow a simpler PIN code to be used for unlocking the Cryptostick and decrypt the Tom container key master material. After 3 false attempts the Cryptostick locks itself down.

**Comment by jaromil** [comment]
Yes indeed, I see the meaningless password prompt popping up. And yes you are right, better to have a switch, more explicit and easier to document in the manual. I'll give this another go soon, looking for a good way to do it that does not add code, eventually removes it from the execution flow.

**Comment by hellekin** [comment]
Didn't test yet...  Wondering though, will that work / did you test with `gpg-agent`?

**Comment by jaromil** [comment]
I postpone this one for after 2.1 release since

1- we did not test it enough yet
2- still does not adopts a flag instead of a file extension as discussed
3- still asks for a fake password prompt going past with a blank return before working as expected
 
Rest assured this feature will go in master at some point, however just not now. Meanwhile, anyone using the `gnupg-delegation` branch is welcome to report experience, observations and fixes.

**Comment by Mindey** [comment]
Works for me :) (the annoyance is a nuisance)

**Comment by saro** [comment]
Hi, any plan for this feature to land in master?

**Comment by jaromil** [comment]
Plan is in place, but due to lack of time there is no progress yet. Pull requests are very welcome, they should correct the current behaviour to adopt a flag instead of extension and possibly not require a blank return.

**Comment by roddhjav** [comment]
I hope you still accept pull request for this...

See:  #244

**Comment by jaromil** [comment]
Definitely still open especially considering the size and detail of your contribution, obviously you did put a lot of time and attention into this: thanks! Just allow a bit of time so that it can be reviewed, the integration with the test suite certainly helps speeding this up. @parazyd is using my old `gnupg-delegation` branch in a project so I'd certainly like to have his opinion, plus would be good to have @alphazo account on this too since she/he started the issue and I believe is using the feature too.

**Comment by roddhjav** [comment]
Of course take your time to review it :) and do not hesitate to ask for more detail or/and changes.

**Comment by jaromil** [comment]
solved by pull-req #244
